src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -174,170 +174,145 @@ public NetworkPeerConnection(NetworkPeer peer, NetworkPeerClient client, IDateTi
+                };
 
-                try
+                this.logger.LogTrace("Sending message: '{0}'", message);
+
+                using (MemoryStream ms = new MemoryStream())
                 {
-                    foreach (SentMessage messageToSend in this.Messages.GetConsumingEnumerable(this.Cancel.Token))
+                    message.ReadWrite(new BitcoinStream(ms, true)
                     {
-                        processing = messageToSend;
-
-                        Payload payload = messageToSend.Payload;
-                        var message = new Message
-                        {
-                            Magic = this.Peer.Network.Magic,
-                            Payload = payload
-                        };
-
-                        this.logger.LogTrace("Sending message: '{0}'", message);
-
-                        using (MemoryStream ms = new MemoryStream())
-                        {
-                            message.ReadWrite(new BitcoinStream(ms, true)
-                            {
-                                ProtocolVersion = this.Peer.Version,
-                                TransactionOptions = this.Peer.SupportedTransactionOptions
-                            });
-
-                            byte[] bytes = ms.ToArray();
-
-                            this.Client.SendAsync(bytes, this.Cancel.Token).GetAwaiter().GetResult();
-                            this.Peer.Counter.AddWritten(bytes.Length);
-                            processing.Completion.SetResult(true);
-                            processing = null;
-                        }
-                    }
+                        ProtocolVersion = this.Peer.Version,
+                        TransactionOptions = this.Peer.SupportedTransactionOptions
+                    });
+
+                    byte[] bytes = ms.ToArray();
+
+                    await this.Client.SendAsync(bytes, this.Cancel.Token).ConfigureAwait(false);
+                    this.Peer.Counter.AddWritten(bytes.Length);
                 }
-                catch (OperationCanceledException)
+            }
+            catch (Exception ex)
+            {
+                if (ex is OperationCanceledException)
                 {
                     this.logger.LogTrace("Sending cancelled.");
                 }
-                catch (Exception ex)
+                else
                 {
                     this.logger.LogTrace("Exception occurred: '{0}'", ex.ToString());
-                    unhandledException = ex;
+
+                    this.Peer.State = NetworkPeerState.Failed;
+                    this.Peer.DisconnectReason = new NetworkPeerDisconnectReason()
+                    {
+                        Reason = "Unexpected exception while sending a message",
+                        Exception = ex
+                    };
                 }
 
-                if (processing != null)
-                    this.Messages.Add(processing);
+                this.Cancel.Cancel();
```

Comment:
``` Cancel.Cancel() ``` I wonder if it might be better with calling it CancellationSource or CancelSrc or something like that so you don't end up with Cancel.Cancel() or maybe even ```Process.Cancel()```?

Reply:
I agree, I will rename it CancellationSource in part 15